File: ports/jagentic-core/src/test/java/org/jagentic/core/conformance/ConformanceTest.java
package org.jagentic.core.conformance;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

import org.junit.jupiter.api.Assumptions;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestFactory;

/**
 * One dynamic test per fixture in {@code spec/conformance/v1/fixtures}. A fixture whose
 * {@code requires} names a capability this runtime lacks is aborted (reported as skipped by JUnit),
 * never passed.
 */
class ConformanceTest {

  @TestFactory
  Stream<DynamicTest> fixtures() {
    List<Path> files = ConformanceHarness.fixtureFiles();
    assertEquals(15, files.size(), "expected the 15 v1 fixtures at " + ConformanceHarness.fixturesDir());
    return files.stream().map(p -> DynamicTest.dynamicTest(p.getFileName().toString(), () -> {
      ConformanceHarness.Outcome o = ConformanceHarness.run(p);
      Assumptions.assumeFalse(o.skipped(), () -> "skip " + o.id() + ": " + o.skipReason());
      assertTrue(o.problems().isEmpty(), () -> "FAIL " + o.id() + "\n  " + String.join("\n  ", o.problems()));
    }));
  }

  @Test
  void comparatorMatchesReferenceRules() {
    Map<String, Object> call = new java.util.HashMap<>(Map.of("tool", "balance", "index", 0, "attempt", 1,
        "args", Map.of("user", "anonymous")));
    call.put("error", null);
    Map<String, Object> actual = Map.of(
        "conversation_id", "c1", "status", "completed", "path", "billing", "reply", "[billing] ok",
        "state", Map.of("turn_count", 2L),
        "tool_calls", List.of(call),
        "events", List.of(Map.of("type", "turn_received"), Map.of("type", "routed"),
            Map.of("type", "tool_called"), Map.of("type", "turn_completed")),
        "error", Map.of("class", "tool"));

    assertTrue(ConformanceHarness.checkExpectation(Map.of(
        "conversation_id", "c1", "status", "completed", "path", "billing", "reply_matches", "^\\[billing\\]",
        "error_class", "tool", "state_includes", Map.of("turn_count", 2),
        "tool_calls", List.of(Map.of("tool", "balance", "index", 0, "args", Map.of("user", "anonymous"))),
        "events_include", List.of("turn_received", "tool_called"),
        "events_exclude", List.of("turn_failed")), actual).isEmpty());

    List<String> problems = ConformanceHarness.checkExpectation(Map.of(
        "status", "failed",
        "tool_calls", List.of(Map.of("tool", "balance", "failed", true)),
        "events_include", List.of("tool_called", "routed"),
        "events_exclude", List.of("routed"),
        "state_includes", Map.of("turn_count", 3)), actual);
    assertEquals(5, problems.size(), problems.toString());
    assertFalse(ConformanceHarness.checkExpectation(Map.of("reply", "other"), actual).isEmpty());
  }
}

File: agentic-clj/test/agentic/conformance_test.clj
(ns agentic.conformance-test
  "Runs the shared v1 fixtures from spec/conformance/v1/fixtures against this runtime. Every fixture
   must pass or be an honest skip; a skip is reported, never counted as a pass."
  (:require [clojure.string :as str]
            [clojure.test :refer [deftest is testing]]
            [agentic.conformance :as conf]))

(defn- justified-skip?
  "A skip is legitimate only when the fixture requires a capability this runtime does not claim
   and the reason names every such capability."
  [fixture reason]
  (let [undeclared (conf/unsupported fixture)]
    (and (seq undeclared)
         (string? reason)
         (every? #(str/includes? reason %) undeclared))))

(deftest shared-fixtures
  (let [files    (conf/fixture-files)
        outcomes (conf/run-all)
        by-id    (into {} (map (juxt :id identity)) outcomes)]
    (is (seq files) (str "no fixtures found at " (conf/fixtures-dir)))
    (is (= (count files) (count outcomes)) "every fixture file in the spec is exercised")
    (doseq [f files
            :let [fixture (conf/read-fixture f)
                  id      (get fixture "id")
                  {:keys [status problems]} (get by-id id)]]
      (testing id
        (case status
          :passed (is true)
          :skipped (do (is (justified-skip? fixture (first problems))
                           (str id " skipped without an undeclared capability named in " (pr-str problems)))
                       (println "SKIP" id (first problems)))
          :failed (is (empty? problems) (str id " " (pr-str problems)))
          (is false (str id " produced no outcome")))))
    (println (conf/report outcomes))))

(deftest comparison-rules
  (testing "events_include is an ordered subsequence, extras allowed"
    (let [actual {"events" [{"type" "turn_received"} {"type" "routed"} {"type" "x"} {"type" "turn_completed"}]}]
      (is (empty? (conf/check-expectation {"events_include" ["turn_received" "turn_completed"]} actual)))
      (is (seq (conf/check-expectation {"events_include" ["turn_completed" "routed"]} actual)))
      (is (seq (conf/check-expectation {"events_exclude" ["routed"]} actual)))))
  (testing "tool_calls compares tool, order and the named keys; failed means an error is recorded"
    (let [actual {"tool_calls" [{"tool" "a" "index" 0 "attempt" 1 "error" {"class" "tool"}}
                                {"tool" "a" "index" 0 "attempt" 2}]}]
      (is (empty? (conf/check-expectation
                   {"tool_calls" [{"tool" "a" "index" 0 "failed" true} {"tool" "a" "attempt" 2}]} actual)))
      (is (seq (conf/check-expectation {"tool_calls" [{"tool" "a"} {"tool" "a" "failed" true}]} actual)))
      (is (seq (conf/check-expectation {"tool_calls" [{"tool" "a"}]} actual)))))
  (testing "state_includes is a subset check, reply_matches a regex search"
    (is (empty? (conf/check-expectation {"state_includes" {"turn_count" 2} "reply_matches" "bal"}
                                        {"state" {"turn_count" 2 "extra" 1} "reply" "your balance"})))
    (is (seq (conf/check-expectation {"reply_matches" "^balance"} {"reply" "your balance"})))))

(deftest unsupported-requirements-skip
  (is (= ["time_travel_ui"] (conf/unsupported {"requires" ["tools" "time_travel_ui"]})))
  (is (= [] (conf/unsupported {"requires" ["tools" "saga"]}))))
